numba/caching.py

```diff
@@ -513,3 +530,56 @@ def _save_data(self, name, cres):
 
     def _dump(self, obj):
         return pickle.dumps(obj, protocol=-1)
+
+    def _reduce(self, cres):
+        """
+        Returns a serialized CompileResult
+        """
+        return cres._reduce()
+
+    def _rebuild(self, target_context, payload):
+        """
+        Returns the unserialized CompileResult
+        """
+        return compiler.CompileResult._rebuild(target_context, *payload)
+
+
+class LibraryCache(FunctionCache):
+    """
+    Extends FunctionCache to provide caching of related CodeLibrary objects.
+    This stores the serialized library in the same cache file as other overload
+    entries for the associated function.
+    """
+    def __init__(self, py_func, identifier):
+        """
+        The *py_func* arg is the associated function being cached.
+        The *identifier* arg is an arbitrary object to identity the feature for
+        which this cache is serving.
+        """
+        self._identifier = identifier
+        super(LibraryCache, self).__init__(py_func)
+
+    def _check_cachable(self, library):
```

Comment:
I agree that it is not good design and hacky.  I was trying to minimize code changes for this patch and do the Cache redesign later.  Since it really bothers you, I will accelerate my plan to redesign the *Cache classes.